File: samples/java/Bingo/Server/Session/src/main/java/systems/zlink/samples/bingo/server/session/sessions/handlers/AuthenticateSessionHandler.java
package systems.zlink.samples.bingo.server.session.sessions.handlers;

import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRouteClient;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkTypedSessionPacketHandler;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

public final class AuthenticateSessionHandler
        implements ZLinkTypedSessionPacketHandler<ZLinkSessionContext, Messages.AuthenticateReq> {
    private final ZLinkRouteClient routes;
    private final ZLinkActorManager actors;

    public AuthenticateSessionHandler(ZLinkRouteClient routes, ZLinkActorManager actors) {
        this.routes = routes;
        this.actors = actors;
    }

    @Override
    public Class<Messages.AuthenticateReq> messageType() {
        return Messages.AuthenticateReq.class;
    }

    @Override
    public CompletionStage<Void> handle(
            ZLinkSessionContext context,
            ZLinkSessionDispatchContext dispatch,
            Messages.AuthenticateReq request) {
        if (request.getAccessToken().isBlank()) {
            throw new IllegalArgumentException("access token is required");
        }
        // --8<-- [start:doc-bingo-session-auth]
        return routes.requestToChannel(
                        SampleNames.ApiChannel,
                        BingoMessages.authenticatePlayerReq(request.getAccessToken()))
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.AuthenticatePlayerRes.class)
                .thenCompose(
                        authenticated -> {
                            requireAuthenticated(authenticated);
                            return actors.getOrCreate(
                                            authenticated.getActorId(), SampleNames.PlayerActorType)
                                    .request(
                                            BingoMessages.ensurePlayerActorReq(
                                                    authenticated.getActorId(),
                                                    authenticated.getDisplayName()))
                                    .submit()
                                    // --8<-- [start:doc-bingo-session-bind]
                                    .thenCompose(
                                            result ->
                                                    context.actors()
                                                            .bind(requireActor(result))
                                                            .thenRun(
                                                                    () ->
                                                                            context.client()
                                                                                    .reply(
                                                                                            BingoMessages
                                                                                                    .authenticateRes(
                                                                                                            authenticated
                                                                                                                    .getActorId(),
                                                                                                            authenticated
                                                                                                                    .getDisplayName()))
                                                                                    .submit()));
                            // --8<-- [end:doc-bingo-session-bind]
                        });
        // --8<-- [end:doc-bingo-session-auth]
    }

    private static void requireAuthenticated(Messages.AuthenticatePlayerRes authenticated) {
        if (!authenticated.getAccepted()
                || authenticated.getActorId().isBlank()
                || authenticated.getDisplayName().isBlank()) {
            throw new IllegalStateException(
                    authenticated.getReason().isBlank()
                            ? "Player authentication failed."
                            : authenticated.getReason());
        }
    }

    private static ActorRef requireActor(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        throw new IllegalStateException("Player actor creation was rejected.");
    }
}

File: samples/kotlin/Bingo/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/session/sessions/handlers/AuthenticateSessionHandler.kt
package systems.zlink.samples.kotlin.bingo.server.session.sessions.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.channels.ZLinkRouteClient
import systems.zlink.framework.kotlin.ZLinkSuspendingTypedSessionPacketHandler
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticatePlayerReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticatePlayerRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticateReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.AuthenticateRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.EnsurePlayerActorReq

class AuthenticateSessionHandler(
    private val routes: ZLinkRouteClient,
    private val actors: ZLinkActorManager,
) : ZLinkSuspendingTypedSessionPacketHandler<ZLinkSessionContext, AuthenticateReq> {
    override fun packetName(): String = "AuthenticateReq"

    override fun messageType(): Class<AuthenticateReq> = AuthenticateReq::class.java

    override suspend fun handle(
        context: ZLinkSessionContext,
        dispatch: ZLinkSessionDispatchContext,
        request: AuthenticateReq,
    ) {
        if (request.accessToken.isBlank()) {
            throw IllegalArgumentException("access token is required")
        }

        // --8<-- [start:doc-bingo-session-auth]
        val authenticated =
            routes
                .requestToChannel(
                    SampleNames.ApiChannel,
                    AuthenticatePlayerReq(request.accessToken),
                )
                .timeout(SampleTimings.RequestTimeout)
                .submit(AuthenticatePlayerRes::class.java)
                .await()
        if (
            !authenticated.accepted ||
                authenticated.actorId.isBlank() ||
                authenticated.displayName.isBlank()
        ) {
            throw IllegalStateException(authenticated.reason ?: "Player authentication failed.")
        }
        val actor =
            actors
                .kotlin()
                .getOrCreate(authenticated.actorId, SampleNames.PlayerActorType)
                .request(EnsurePlayerActorReq(authenticated.actorId, authenticated.displayName))
                .await()
        // --8<-- [end:doc-bingo-session-auth]
        // --8<-- [start:doc-bingo-session-bind]
        context.actors().bind(requireActor(actor)).await()
        context
            .client()
            .reply(AuthenticateRes(authenticated.actorId, authenticated.displayName))
            .submit()
        // --8<-- [end:doc-bingo-session-bind]
    }

    private fun requireActor(result: ZLinkActorCreateResult): ActorRef =
        when (result) {
            is ZLinkActorCreateResult.Existing -> result.actor
            is ZLinkActorCreateResult.Created -> result.actor
            is ZLinkActorCreateResult.Rejected ->
                throw IllegalStateException("Player actor creation was rejected.")
        }
}
